jni/src/test/java/com/horizen/certnative/NaiveThresholdSignatureWKeyRotationProofTest.java

```diff
@@ -0,0 +1,276 @@
+        Random r = new Random();
+
+        scId = FieldElement.createRandom();
+        prevScId = FieldElement.createRandom();
+
+        backwardTransferCount = r.nextInt(backwardTransferCount + 1);
+        // Create dummy Backward Transfers
+        List<BackwardTransfer> btList = new ArrayList<>();
+        for(int i = 0; i < backwardTransferCount; i++)
+            btList.add(BackwardTransfer.getRandom(r));
+
+        List<BackwardTransfer> prevBtList = new ArrayList<>();
+        for(int i = 0; i < backwardTransferCount; i++)
+            prevBtList.add(BackwardTransfer.getRandom(r));
+
+        // Compute keys and signatures
+        List<SchnorrKeyPair> signingKeyPairList = new ArrayList<>();
+        List<SchnorrPublicKey> signingPublicKeyList = new ArrayList<>();
+        List<SchnorrPublicKey> masterPublicKeyList = new ArrayList<>();
+        List<SchnorrPublicKey> updatedSigningKeyList = new ArrayList<>();
+        List<SchnorrPublicKey> updatedMasterKeyList = new ArrayList<>();
+        List<SchnorrSignature> updatedSigningKeysSkSignaturesList = new ArrayList<>();
+        List<SchnorrSignature> updatedSigningKeysMkSignaturesList = new ArrayList<>();
+        List<SchnorrSignature> updatedMasterKeysSkSignaturesList = new ArrayList<>();
+        List<SchnorrSignature> updatedMasterKeysMkSignaturesList = new ArrayList<>();
+        List<SchnorrSignature> signatureList = new ArrayList<>();
+
+        for (int i = 0; i<keyCount; i++) {
+            SchnorrKeyPair keyPair = SchnorrKeyPair.generate();
+
+            assertNotNull("Key pair generation was unsuccessful.", keyPair);
+            assertTrue("Public key verification failed.", keyPair.getPublicKey().verifyKey());
+
+            signingKeyPairList.add(keyPair);
+            signingPublicKeyList.add(keyPair.getPublicKey());
+            updatedSigningKeyList.add(keyPair.getPublicKey());
+
+            SchnorrKeyPair masterKeyPair = SchnorrKeyPair.generate();
+
+            assertNotNull("Key pair generation was unsuccessful.", masterKeyPair);
+            assertTrue("Public key verification failed.", masterKeyPair.getPublicKey().verifyKey());
+
+            masterPublicKeyList.add(masterKeyPair.getPublicKey());
+            updatedMasterKeyList.add(masterKeyPair.getPublicKey());
+            masterKeyPair.getSecretKey().freeSecretKey();
+        }
+
+        keysSignaturesList = new ValidatorKeysUpdatesList(
+            signingPublicKeyList,
+            masterPublicKeyList,
+            updatedSigningKeyList,
+            updatedMasterKeyList,
+            updatedSigningKeysSkSignaturesList,
+            updatedSigningKeysMkSignaturesList,
+            updatedMasterKeysSkSignaturesList,
+            updatedMasterKeysMkSignaturesList
+        );
+        validatorsKeysRoot = keysSignaturesList.getKeysRootHash(keyCount);
+        prevValidatorsKeysRoot = keysSignaturesList.getUpdatedKeysRootHash(keyCount);
+        if (usePrevCert) {
```

Comment:
Shoudn't the `genesisKeyRootHash` be always correspond to `validatorsKeyRoot` rather than emplopying the root of updated keys in case `usePrevCert` is true?

Reply:
I actually checked your modifications but it seems to me that you wrongly initialized the `genesisKeyRootHash` value: indeed, you initialize it with the merkle root of the updated set of keys rather than the current set of keys; this should lead to a failure of the circuit if the first certificate is updating a key: indeed, the circuit will check that `genesisKeyRootHash` is equivalent to the merkle root of the current set of keys, which will be different from the merkle root of the updated set of keys. This potential error would have been spot by testing also the case when keys are updated, so maybe it would be helpful to add to the java tests also a test case where the keys are updated (maybe you can simply add an input flag to `testCreateVerifyRandomProof` to avoid duplicating the test

Reply:
Done.